README
Computer Science from the Bottom Up
-----------------------------------

You should be able to make the course materials by typing

make html (output in ./html directory)

or

make pdf (output in csbu.pdf)

Included for simplicity are the docbook dtds, xsl stylesheets and fop
distribution.  You need some other things for this to work out of the
box.  If you have the Debian packages

* java (1.8 ... 1.7 seems to have some issues with saxon; see
    https://sourceforge.net/p/docbook/bugs/1373/ )
* libxml2-utils
* jing
* transfig (for fig2dev)

you will probably be ok.

I am always interested in contribution and feedback so please feel
free to email

Ian Wienand
ian@wienand.org

---

Notes

 2016-01-13 : fop-2.0/lib/batik-all-svn-trunk.jar from [1] is included
 to work around [2]

[1] https://svn.apache.org/repos/asf/xmlgraphics/fop/trunk/lib/batik-all-svn-trunk.jar
[2] https://issues.apache.org/jira/browse/FOP-2489

docbook-5.0/ChangeLog
2008-02-06 14:01  nwalsh

	* spec/docbook.xml: Fix document draft status

2008-02-06 13:57  nwalsh

	* howto/howto.xml: Fixed pubdate

2008-02-06 13:53  nwalsh

	* spec/docbook.xml: Fixed pubdate

2008-02-06 13:52  nwalsh

	* dbforms/dbforms.rnc, dbforms/htmlform.rnc, defguide/defguide.rnc,
	  docbook/docbook.rnc, docbookxi/docbookxi.rnc,
	  docbookxi/xinclude.rnc, spec/docbook.xml, src/admonitions.rnc,
	  src/annotations.rnc, src/bibliography.rnc, src/callouts.rnc,
	  src/calstbl.rnc, src/core.rnc, src/docbook.rnc, src/ebnf.rnc,
	  src/error.rnc, src/glossary.rnc, src/gui.rnc, src/hier.rnc,
	  src/htmltbl.rnc, src/index.rnc, src/keyboard.rnc, src/markup.rnc,
	  src/math.rnc, src/mathml.rnc, src/msgset.rnc, src/os.rnc,
	  src/pool.rnc, src/product.rnc, src/programming.rnc,
	  src/qandaset.rnc, src/refentry.rnc, src/refsect1.rnc,
	  src/sect1.rnc, src/svg.rnc, src/tasks.rnc, src/technical.rnc,
	  src/toc.rnc, src/xlink.rnc: DocBook V5.0 released

2008-01-23 00:55  rlhamilton

	* howto/howto.xml: Copy edit for spelling, typos and minor
	  grammatical
	  glitches. Updated the link to the xsl2 stylesheet
	  snapshot.

2007-12-16 17:03  nwalsh

	* tests/local/dtd.xml: Tweaked to test refsect2 bugs

2007-12-14 19:17  mzjn

	* src/pool.rnc: Removed "other than a corporation" from refpurpose
	  for orgname.

2007-12-07 17:55  kosek

	* howto/howto.xml: Added link to a new superentity file.

2007-11-24 17:24  mzjn

	* howto/howto.xml: Fixed typo.

2007-11-24 16:46  mzjn

	* howto/howto.xml: Some punctuation, spelling, and wording fixes.

2007-11-17 14:33  mzjn

	* howto/howto.xml: Removed obsolete paragraph saying that profiling
	  stylesheets are needed for HTML Help,
	  JavaHelp, or Eclipse.
	  Added info about the warning message that is output if the
	  stylesheets are unable to
	  add @xml:base.
	  Changed Saxon 8 to Saxon 9.

2007-11-10 16:50  kosek

	* howto/howto.xml: Add instructions for using JNVDL

2007-11-10 09:31  kosek

	* howto/howto.xml: Prepare for 5.0 release

2007-10-28 22:14  kosek

	* howto/howto.xml: Validation with XIncludes, new links

2007-10-28 14:19  kosek

	* howto/howto.xml: Fixed typo

2007-10-28 10:38  kosek

	* howto/TODO, howto/howto.css, howto/howto.xml: Small improvements,
	  updated to CR7

2007-10-28 10:20  kosek

	* howto/howto.xml: Information about referencing DocBook schema

2007-10-27 21:48  kosek

	* howto/TODO, howto/howto.xml: Some additions for a new release

2007-10-02 02:20  nwalsh

	* src/callouts.rnc: Fix typo in documentation for the units
	  enumeration

2007-09-28 19:06  nwalsh

	* spec/docbook.xml: Updated for 5.0CR7

2007-09-28 18:50  nwalsh

	* dbforms/dbforms.rnc, dbforms/htmlform.rnc, defguide/defguide.rnc,
	  docbook/docbook.rnc, docbookxi/docbookxi.rnc,
	  docbookxi/xinclude.rnc, src/admonitions.rnc, src/annotations.rnc,
	  src/bibliography.rnc, src/callouts.rnc, src/calstbl.rnc,
	  src/core.rnc, src/docbook.rnc, src/ebnf.rnc, src/error.rnc,
	  src/glossary.rnc, src/gui.rnc, src/hier.rnc, src/htmltbl.rnc,
	  src/index.rnc, src/keyboard.rnc, src/markup.rnc, src/math.rnc,
	  src/mathml.rnc, src/msgset.rnc, src/os.rnc, src/pool.rnc,
	  src/product.rnc, src/programming.rnc, src/qandaset.rnc,
	  src/refentry.rnc, src/refsect1.rnc, src/sect1.rnc, src/svg.rnc,
	  src/tasks.rnc, src/technical.rnc, src/toc.rnc, src/xlink.rnc:
	  DocBook V5.0CR7 released

2007-09-28 18:34  nwalsh

	* src/admonitions.rnc, src/core.rnc, src/docbook.rnc,
	  src/error.rnc, src/gui.rnc, src/keyboard.rnc, src/markup.rnc,
	  src/os.rnc, src/pool.rnc, src/product.rnc, src/programming.rnc,
	  src/technical.rnc: Refactor to make the work of the Publishing
	  Subcommitee easier

2007-09-27 18:09  nwalsh

	* spec/docbook.xml: Updated revision history

2007-09-27 14:03  nwalsh

	* dbforms/dbforms.rnc, dbforms/htmlform.rnc, defguide/defguide.rnc,
	  docbook/docbook.rnc, docbookxi/docbookxi.rnc,
	  docbookxi/xinclude.rnc, spec/docbook.xml, src/annotations.rnc,
	  src/bibliography.rnc, src/callouts.rnc, src/calstbl.rnc,
	  src/docbook.rnc, src/ebnf.rnc, src/glossary.rnc, src/hier.rnc,
	  src/htmltbl.rnc, src/index.rnc, src/math.rnc, src/mathml.rnc,
	  src/msgset.rnc, src/pool.rnc, src/qandaset.rnc, src/refentry.rnc,
	  src/refsect1.rnc, src/sect1.rnc, src/svg.rnc, src/tasks.rnc,
	  src/toc.rnc, src/xlink.rnc: Version 5.0CR6 released

2007-09-27 13:11  nwalsh

	* src/pool.rnc: Fix RFE 1759782: Allow uri anywhere email occurs

2007-09-27 12:55  nwalsh

	* src/hier.rnc, src/pool.rnc: Fix (part of) RFE 1784312: Allow book
	  to be completely empty

2007-09-27 12:52  nwalsh

	* src/pool.rnc: Fix (part of) RFE 1784312: Allow personblurb and
	  titleabbrev in biblioentry (and other bibliographic contexts)

2007-09-27 12:45  nwalsh

	* src/mathml.rnc: Fix RFE 1795884: Allow MathML in inlineequation

2007-09-27 12:43  nwalsh

	* src/pool.rnc: Fix RFE 1800916: Allow keycap etc. in userinput

2007-08-30 01:45  xmldoc

	* howto/stylesheets/fo-ns.xsl, howto/stylesheets/fo.xsl,
	  howto/stylesheets/html-ns.xsl, howto/stylesheets/html.xsl:
	  Updated stylesheets to display data on titlepage from othercredit
	  content (not just in author content). If there is an otherclass
	  attribute, the value of the is displayed. It really should also
	  look for the value of the class attribute, and use that unless it
	  is "other" (in which case it uses the otherclass value instead),
	  but that change will have to wait til later...

2007-08-30 01:41  xmldoc

	* howto/howto.xml: Created dbxsl-ns section and added myself to
	  list of contributors.

2007-08-29 05:25  xmldoc

	* howto/Makefile, howto/stylesheets/fo-ns.xsl,
	  howto/stylesheets/html-ns.xsl: Added customization layers for
	  XSL-NS stylesheets, and set up
	  added targets to makefile for generating output using those.

2007-08-02 20:07  kosek

	* dbits/dbits.rnc: Refactored to use new pattern introduced in CR5

2007-07-18 20:43  nwalsh

	* src/pool.rnc: Move Bob's olink text up to the schema source
	  location

2007-07-12 13:38  nwalsh


fop-2.0/fop.js
//   Licensed to the Apache Software Foundation (ASF) under one or more
//   contributor license agreements.  See the NOTICE file distributed with
//   this work for additional information regarding copyright ownership.
//   The ASF licenses this file to You under the Apache License, Version 2.0
//   (the "License"); you may not use this file except in compliance with
//   the License.  You may obtain a copy of the License at
//
//       http://www.apache.org/licenses/LICENSE-2.0
//
//   Unless required by applicable law or agreed to in writing, software
//   distributed under the License is distributed on an "AS IS" BASIS,
//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//   See the License for the specific language governing permissions and
//   limitations under the License.
//   $Id: fop.js 1390331 2012-09-26 08:50:07Z mehdi $ -->

// jscript to run FOP, adapted from the Jakarta-Ant project.

// rpm_mode is irrelevant on Windows
// var rpm_mode=true;
var fop_exec_args = "";
var no_config=false;
var fop_exec_debug=false;
var debug=false;
var keep_open=false;
var show_help=false;

var config_wanted = new Array("FOP_HOME", "CLASSPATH", "FOP_HYPHENATION_PATH", "FOP_OPTS", "JAVA_OPTS", "LOGCHOICE", "LOGLEVEL");

// parse command-line arguments
function read_args() {
  var args = WScript.Arguments;
  var named = new ActiveXObject("Scripting.Dictionary");
  for (i = 0; i < args.length; i++) {
	switch(args(i)) {
	case "--debug":
	  debug=true;
	  break;
	case "--execdebug":
	  fop_exec_debug=true;
	  break;
	case "--keepopen":
	  keep_open=true;
	  break;
	case "--noconfig":
	  no_config=true;
	  break;
	case "-h":
	case "--help":
	case "--h":
	case "-help":
	  show_help=true;
	  // fop_exec_args=fop_exec_args + " -h";
	  break;
	default:
	  fop_exec_args=fop_exec_args + " " + args(i);
	}
  }
  if (debug) {
	WScript.Echo("debug: " + debug);
	WScript.Echo("execdebug: " + fop_exec_debug);
 	WScript.Echo("keepopen: " + keep_open);
 	WScript.Echo("noconfig: " + no_config);
	WScript.Echo("help: " + show_help);
 	WScript.Echo("fop arguments: " + fop_exec_args);
  }
}

var help_text="Usage:\n"
  + WScript.ScriptFullName + " [script options] [FOP options]\n"
  + "Script Options:\n"
  + "  --help, -h             print this message and FOP help\n"
  + "  --debug                print debugging information for this launch script\n"
  + "  --execdebug            print FOP exec line generated by this launch script\n"
  + "  --keepopen             keep FOP's command window open\n"
  + "  --noconfig             suppress reading of configuration file and registry";

function read_environment() {
  for (i in config_wanted) {
	if (!config.Exists(config_wanted[i])) {
		var env_var_string = "%" + config_wanted[i] + "%";
		var env_var = shell.ExpandEnvironmentStrings(env_var_string);
		if (env_var != "" && env_var != env_var_string) {
		  config.Add(config_wanted[i], env_var);
		  if (debug) {
			WScript.Echo(config_wanted[i] + " env: "
						 + config.Item(config_wanted[i]));
		  }
		}
	}
  }
}

function read_desktop(dtname) {
  if (fs.FolderExists(dtname)) {
	var fopname = fs.GetFolder(dtname).ParentFolder.Path
	  + "\\Application Data\\Fop";
	if (fs.FolderExists(fopname)) {
	  var fop_conf_name = fs.GetFolder(fopname).Path + "\\fop.conf";
	  if (fs.FileExists(fop_conf_name)) {
		// source fop_conf_file
		var conf_file = fs.openTextFile(fs.GetFile(fop_conf_name));
		var conf_lines = new ActiveXObject("Scripting.Dictionary");
		while (!conf_file.AtEndOfStream) {
		  var line = conf_file.ReadLine();
		  var m = line.match(/(.+?)=(.+)/);
		  if (m != null) {
			conf_lines.Add(m[1], m[2]);
		  }
		}
		for (j in config_wanted) {
		  if (!config.Exists(config_wanted[j])
			  && conf_lines.Exists(config_wanted[j])) {
			config.Add(config_wanted[j], conf_lines.Item(config_wanted[j]));
			if (debug) {
			  WScript.Echo(config_wanted[j] + " " + dts[i] + ": "
						   + config.Item(config_wanted[i]));
			}
		  }
		}
	  }
	}
  }
}

function read_registry(section) {
  for (j in config_wanted) {
	if (!config.Exists(config_wanted[j])) {
	  var reg_var;
	  try {
		reg_var = shell.RegRead(section + "\\Software\\Fop\\"
								+ config_wanted[j]);
		config.Add(config_wanted[j], reg_var);
		if (debug) {
		  WScript.Echo(config_wanted[j] + " " + rks[i] + ": "
					   + config.Item(config_wanted[j]));
		}
	  } catch(e) {}
	}
  }
}

// construct FOP_HOME from the script folder
function get_fop_home() {
  if (!config.Exists("FOP_HOME")
	  || !fs.FolderExists(config.Item("FOP_HOME"))) {
	var fop_home = WScript.ScriptFullName;
	fop_home = fop_home.substring(0, fop_home.length
								  - WScript.ScriptName.length - 1);
	if (config.Exists("FOP_HOME")) {
	  config.Remove("FOP_HOME");
	}
	config.Add("FOP_HOME", fop_home);
	if (debug) {
	  WScript.Echo("FOP_HOME dyn: " + config.Item("FOP_HOME"));
	}
  }
}

function get_java_cmd() {
  var java_home = shell.ExpandEnvironmentStrings("%JAVA_HOME%");
  javacmd = "java";
  if (java_home != "" && typeof(java_home) != "undefined"
	  && fs.FolderExists(java_home)) {
	var javacmd_candidate = java_home + "\\bin\\java.exe";
	if (fs.FileExists(javacmd_candidate)) {
	  javacmd = javacmd_candidate;
	}
  }
  if (debug) {
	WScript.Echo("java command: " + javacmd);
  }
}

function get_local_classpath() {
  if (config.Exists("CLASSPATH")) {
	local_classpath = config.Item("CLASSPATH");
	if (debug) {
	  WScript.Echo("local classpath: " + local_classpath);
	}
  }

  // add fop.jar, fop-sandbox and fop-hyph.jar, which reside in $FOP_HOME/build
  var lcp = local_classpath;
  local_classpath = config.Item("FOP_HOME") + "\\build\\fop.jar;"
	+ config.Item("FOP_HOME") + "\\build\\fop-sandbox.jar;"
	+ config.Item("FOP_HOME") + "\\build\\fop-hyph.jar";
  if (lcp != "") {
	local_classpath += ";" + lcp;
  }
  if (debug) {
	WScript.Echo("local classpath: " + local_classpath);
  }

  // add in the dependency .jar files, which reside in $FOP_HOME/lib
  var libdir_name = config.Item("FOP_HOME") + "\\lib";
  var dirlibs;
  if (fs.FolderExists(libdir_name)) {
	dirlibs = fs.GetFolder(libdir_name).Files;
	var e = new Enumerator(dirlibs);
	for (; !e.atEnd(); e.moveNext()) {
	  if (e.item().Name.match("\.jar$")) {
		local_classpath = libdir_name + "\\" + e.item().Name + ";" + local_classpath;
	  }
	}
	if (debug) {
	  WScript.Echo("local classpath: " + local_classpath);
	}
  }

  // add in user-defined hyphenation JARs
  if (config.Exists("FOP_HYPHENATION_PATH")) {
	local_classpath += ";" + config.Item("FOP_HYPHENATION_PATH");
	if (debug) {
	  WScript.Echo("local classpath: " + local_classpath);
	}
  }
}

// Execute fop via shell.Exec
function fop_exec() {
  var fop_exec_command = "\"" + javacmd + "\" "
	+ (config.Exists("JAVA_OPTS")?config.Item("JAVA_OPTS") + " ":"")
	+ (config.Exists("LOGCHOICE")?config.Item("LOGCHOICE") + " ":"")
	+ (config.Exists("LOGLEVEL")?config.Item("LOGLEVEL") + " ":"")
	+ "-classpath \"" + local_classpath + "\" "
	+ (config.Exists("FOP_OPTS")?config.Item("FOP_OPTS"):"")
	+ "org.apache.fop.cli.Main " + fop_exec_args;
  if (debug || fop_exec_debug) {
	WScript.Echo(fop_exec_command);
  }

  var fop_run = shell.Exec(fop_exec_command);
  while (true) {
	while (!fop_run.StdOut.AtEndOfStream) {
	  WScript.Echo(fop_run.StdOut.ReadLine());
	}
	while (!fop_run.StdErr.AtEndOfStream) {
	  WScript.Echo(fop_run.StdErr.ReadLine());
	}
	if (fop_run.Status == 1) {
	  break;
	}
	WScript.Sleep(100);
  }
  if (debug) {
	WScript.Echo("exit status: " + fop_run.ExitCode);
  }
}

// Execute fop via shell.Run
function fop_run() {
  var fop_exec_command = "cmd /" + (keep_open?"K":"C") + " \""
	+ "\"" + javacmd + "\" " 
	+ (config.Exists("JAVA_OPTS")?config.Item("JAVA_OPTS") + " ":"")
	+ (config.Exists("LOGCHOICE")?config.Item("LOGCHOICE") + " ":"")
	+ (config.Exists("LOGLEVEL")?config.Item("LOGLEVEL") + " ":"")
	+ "-classpath \"" + local_classpath + "\" "
	+ (config.Exists("FOP_OPTS")?config.Item("FOP_OPTS") + " ":"")
	+ "org.apache.fop.cli.Main " + fop_exec_args + "\"";
  if (debug || fop_exec_debug) {
	WScript.Echo(fop_exec_command);
  }
  var exit_code = shell.Run(fop_exec_command, 1, 1);
  if (debug) {
	WScript.Echo("exit status: " + exit_code);
  } else {
	if (exit_code != 0) {
	  WScript.Echo("A FOP error occurred (FOP exit status: " + exit_code + ")\n"
	               + "Use option --keepopen to see FOP's output\n"
                       + "(that is two dashes)");
	}
  }
}

function get_log_choice() {
  // The default commons logger for JDK1.4 is JDK1.4Logger.
  // To use a different logger, uncomment the one desired below
  if (!config.Exists("LOGCHOICE")) {
	// config.Add("LOGCHOICE","\"-Dorg.apache.commons.logging.Log=org.apache.commons.logging.impl.NoOpLog\"");
	// config.Add("LOGCHOICE","\"-Dorg.apache.commons.logging.Log=org.apache.commons.logging.impl.SimpleLog\"");
	// config.Add("LOGCHOICE","\"-Dorg.apache.commons.logging.Log=org.apache.commons.logging.impl.Log4JLogger\"");
	if (debug && config.Exists("LOGCHOICE")) {
	  WScript.Echo("LOGCHOICE script: " + config.Item("LOGCHOICE"));
	}
  }
}

function get_log_level() {
  // Logging levels
  // Below option is only if you are using SimpleLog instead of the default JDK1.4 Logger.
  // To set logging levels for JDK 1.4 Logger, edit the %JAVA_HOME%/JRE/LIB/logging.properties 
  // file instead.
  // Possible SimpleLog values:  "trace", "debug", "info" (default), "warn", "error", or "fatal".
  if (!config.Exists("LOGLEVEL")) {
	// config.Add("LOGLEVEL","\"-Dorg.apache.commons.logging.simplelog.defaultlog=INFO\"");
	if (debug && config.Exists("LOGLEVEL")) {
	  WScript.Echo("LOGLEVEL script: " + config.Item("LOGLEVEL"));
	}
  }
}

var shell = WScript.CreateObject("WScript.Shell");
var fs = WScript.CreateObject("Scripting.FileSystemObject");

// configuration
var config = new ActiveXObject("Scripting.Dictionary");

read_args();
read_environment();
if (!no_config) {
  // read user and system-wide fop configurations
  var spec = shell.SpecialFolders;
  var dts = new Array("Desktop", "AllUsersDesktop");
  for (i in dts) {
	read_desktop(spec(dts[i]));
  }
  // read user and system-wide registry
  var rks = new Array("HKCU", "HKLM");
  for (i in rks) {
	read_registry(rks[i]);
  }
}

get_fop_home();
get_log_choice();
get_log_level();
var javacmd = "";
get_java_cmd();
var local_classpath = "";
get_local_classpath();

// Show script help if requested
if (show_help) {
  // fop_exec_args = "";
  keep_open = true;
  WScript.Echo(help_text);
}

// fop_exec();
fop_run();
